### PR TITLE
json based convo agent - for discussion

### PR DESCRIPTION
**Pull request by jacobrosenthal**
This is a json based convo fork ive been playing with. It takes json only from chatgpt, defining its api via typescript examples. This is less token verbose and ive found it to be more reliable than the upstream one as constraining to a json array for actions really seems to give it less ways to screw up

I think the COT prompting with thought: is really necessary, and with the current agent interface its pretty ackwardly written. Not particularly extendable. 

Its a bit slower, because I dont use the stop, so it has to receive all the tokens. Maybe theres a way to reliably detect like the comma between two objects

For discussion, maybe someone has a way to clean it up or just abandon the idea that its extendable

~All the other commits are just Prs I needed to build this on top of out of tree, just the last commit is the agent when everything else clears.~
related to my request in https://github.com/hwchase17/langchainjs/pull/349

### CODE REVIEW

**Comment by jacobrosenthal** [comment]
Much of the time chatgpt responds with
```
langchain-examples:start:   "action_input": "pow(Harry Styles age, 0.23)",
```
but occasionally it uses
```
"action_input": `${Math.pow((new Date().getFullYear() - 1994), 0.23)}`
```
which kills the json parsing.. whats a good escape library for json parsing? Or any ideas on prompt engineering? If you tell it to never use ` in the prompt itll fail with
```
langchain-examples:start:   "action_input": "round(" + Math.pow(age, 0.23) + ")",
```

**Comment by folke** [comment]
This is really funny. I also started using TypeScript types and initially used JSON. But as you also discovered, it's hard to parse invalid json.
I eventually just used the same types, but said it had to respond with YAML instead. It's much better in writing yaml in my experience.

**Comment by jacobrosenthal** [comment]
Oh interesting. Or your agent? ;)

**Comment by jacobrosenthal** [comment]
Pr. Pull request

**Comment by jacobrosenthal** [comment]
I tried some yaml but didnt have a lot of luck, it doesnt feel constrained enough and it likes to not follow the schema. Not using certain fields randomly. This json is still working best for me, assuming im not using the math tool and no backticks show up.
